### PR TITLE
Support for isolated and async deinit

### PR DESCRIPTION
**Pull request by nickolas-pohilets**
See also https://github.com/apple/swift/pull/60057

### CODE REVIEW

File: Sources/SwiftParser/Declarations.swift
Comment (ahoppen) [inline Sources/SwiftParser/Declarations.swift:1154]: Incorrect  indentation. Could you run `./format.py`?

File: Sources/SwiftParser/Specifiers.swift
Comment (ahoppen) [inline Sources/SwiftParser/Specifiers.swift:445]: Could you order them alphabetically? It just makes it easier to read IMO. Same in the switches below

File: Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift
Comment (ahoppen) [inline Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift:173]: Out of curiosity: What motivated this change?

File: Sources/SwiftParser/Specifiers.swift
Comment (ahoppen) [inline Sources/SwiftParser/Specifiers.swift:421]: Oh, I just noticed that `MisspelledAsyncSpecifiers` is inconsistently named with `MisspelledThrowsTokenKinds`. I’m fixing that in https://github.com/apple/swift-syntax/pull/1767. You’ll probably need to rebase once that PR is merged.

File: Sources/SwiftParser/Specifiers.swift
Comment (ahoppen) [inline Sources/SwiftParser/Specifiers.swift:474]: Also, is it really worth having this entire dance of re-arranging the tokens? It seems overly complicated to me and possibly doesn’t provide too much value. AFAICT the only place where it’s used is when parsing a function signature after `deinit` and honestly, if you put a function signature after `deinit`, put a `throws` into it and then also do that in the wrong order, I’m not sure if that’s really a recovery case we need to add a >100-line recovery path for. 

If we do want to keep it, here are some more specific comments

I think it would be worth adding a doc-comment here saying that `throwsSpecifier` will be stored as an unexpected token because `RawDeinitEffectSpecifiersSyntax` doesn’t actually have an  throws specifier.

Could you also mark it as `@_spi(RawSyntax)`?

File: Sources/SwiftParser/Specifiers.swift
Comment (ahoppen) [inline Sources/SwiftParser/Specifiers.swift:478]: Is `?.raw.as(RawTokenSyntax.self)` needed for anything? AFAICT it’s a no-op.

File: Sources/SwiftParser/Specifiers.swift
Comment (ahoppen) [inline Sources/SwiftParser/Specifiers.swift:646]: This also seems overly complicated. Can’t we just eat consume any misplaced throws specifiers into `unexpectedBeforeAsync` and then just do `self.consume(if: .async)`?

Alternatively, could we keep your hand-written initializer for `RawDeinitEffectSpecifiersSyntax`, make it conform to `RawEffectSpecifiersTrait` and use `parseEffectSpecifiers`?

Comment (nickolas-pohilets) [reply Sources/SwiftParser/Specifiers.swift:646]: I was trying to follow the logic of `parseEffectSpecifiers()` and handle the same recovery cases, including misspelled and duplicated `async`. If I understood you correctly, this simplification would not parse such cases, or would it?

I've tried the latter, but it consumes deinit name into `unexpectedBeforeAsyncSpecifier`.

File: Sources/SwiftParser/Specifiers.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Specifiers.swift:474]: I think this could be simplified if I don't reuse entire `parseFunctionSignature()`, but only smaller functions, like `parseParameterClause()`. I can pass dummy `inout RawFunctionEffectSpecifiersSyntax?` to `parseFunctionReturnClause()` and discard the value. Or don't attempt to parse output part at all. WDYT?

File: Sources/SwiftParser/Declarations.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Declarations.swift:1154]: Sure. Is it ok to run it once as a separate commit, or should I patch every commit?

File: Sources/SwiftParser/Specifiers.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Specifiers.swift:445]: Done

File: Sources/SwiftParser/Specifiers.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Specifiers.swift:421]: Updated

File: Sources/SwiftParser/Specifiers.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Specifiers.swift]: I don't see any tests for deinit with output type, so I guess dropping support for it should be ok. I've pushed an update, please take a look.

File: Sources/SwiftParser/Specifiers.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Specifiers.swift:580]: I would then combine all the loops into one.

Comment (nickolas-pohilets) [reply Sources/SwiftParser/Specifiers.swift]: I've pushed an update, please take a look

File: Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift
Comment (ahoppen) [inline Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift:810]: Can’t you use the `ThrowsEffectSpecifier` token spec set here instead of re-defining the list?

File: Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift
Comment (ahoppen) [inline Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift:831]: `isOfSameKind` is a little misleading here. I would just call it: `isAsyncEffectSpecifier` (and maybe also rename `asThrowingToken` -> `asThrowsEffectSpecifier` for consistency).

File: Sources/SwiftParser/Declarations.swift
Comment (ahoppen) [inline Sources/SwiftParser/Declarations.swift:1136]: I know that you just copied this code but it can be slightly simplified:

```suggestion
    if effectSpecifiers == nil && self.at(TokenSpec(.leftParen, allowAtStartOfLine: false)) {
```

File: Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift
Comment (ahoppen) [reply Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift:173]: Ah, I see. Could you update the doc comment to match the new behavior?

File: Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift
Comment (ahoppen) [inline Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift:805]: You’re missing

```swift
if shouldSkip(node) {
  return .skipChildren
}
```

at the start of the `visit` function. Unfortunately, I don’t know of a good way to automatically add it.

File: Sources/SwiftParser/Declarations.swift
Comment (ahoppen) [inline Sources/SwiftParser/Declarations.swift:1139]: If you’re changing this from `parseFunctionSignature` to `parseParameterClause`, I think we aren’t recovering the following as nicely anymore. Can’t we continue to use `parseFunctionSignature` here anymore? 

```swift
deinit() -> Int {}
```

And also: Could you add a test case for a `deinit` with a return value?

File: Tests/SwiftParserTest/translated/InitDeinitTests.swift
Comment (ahoppen) [inline Tests/SwiftParserTest/translated/InitDeinitTests.swift:789]: This is expected for now. `async` is parsed as a modifier like `public` because you can have an `async var x: Int`.

**Review by ahoppen — COMMENTED** [review]
This looks really good. I’ve got a few minor comments but all of them should be fairly minor.

File: Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift:810]: Done

File: Sources/SwiftParser/Declarations.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Declarations.swift:1136]: Applied

File: Tests/SwiftParserTest/translated/InitDeinitTests.swift
Comment (nickolas-pohilets) [reply Tests/SwiftParserTest/translated/InitDeinitTests.swift:789]: Updated comment

File: Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift:805]: Done

File: Sources/SwiftParser/Declarations.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Declarations.swift:1139]: `parseFunctionSignature()` uses `parseFunctionEffectSpecifiers()` which we don't want. I don't think it is worth to abstract `parseFunctionSignature()` over `parseFunctionEffectSpecifiers()` vs `parseDeinitEffectSpecifiers()`. It would be simpler to use `parseParameterClause()` + `parseDeinitEffectSpecifiers()` + `parseFunctionReturnClause()` directly from `parseDeinitializerDeclaration()`.

As a bonus it would allow recovering from `deinit -> Int {}` which `parseFunctionSignature()` cannot.

Note that using `RawDeinitEffectSpecifiersSyntax` with`parseFunctionReturnClause()` still requires conformance to  `RawEffectSpecifiersTrait` - that's still a bit ugly, but at least now without any logic.

Comment (nickolas-pohilets) [reply Sources/SwiftParser/Declarations.swift:1139]: Updated, please take a look

File: Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift:173]: Updated. Please take a look

File: Sources/SwiftParser/Declarations.swift
Comment (ahoppen) [inline Sources/SwiftParser/Declarations.swift:1148]: `self.at(.arrow)` implies `self.canRecoverTo(TokenSpec(.arrow, recoveryPrecedence: .exprKeyword)) != nil` so I think you can just remove `self.at(.arrow)`.

File: Sources/SwiftParser/Declarations.swift
Comment (ahoppen) [reply Sources/SwiftParser/Declarations.swift:1139]: Oh, nice. That looks good to me.

File: Sources/SwiftParser/Specifiers.swift
Comment (ahoppen) [inline Sources/SwiftParser/Specifiers.swift:486]: We don’t have any precedence for empty `TokenSpecSet` and `canRecoveryTo` has a precondition that the spec set is not empty. 

Could you change 

https://github.com/apple/swift-syntax/blob/8a759449c646c72b9392b151236204a5fad03831/Sources/SwiftParser/Recovery.swift#L133

to 

```swift
if specSet.allCases.isEmpty {
  return nil
}
```

File: Sources/SwiftParser/Specifiers.swift
Comment (ahoppen) [inline Sources/SwiftParser/Specifiers.swift:511]: Could you change this to a `precondition`?

https://github.com/apple/swift-syntax/blob/8a759449c646c72b9392b151236204a5fad03831/Sources/SwiftSyntax/Assert.swift#L13-L22

File: Sources/SwiftParser/Specifiers.swift
Comment (ahoppen) [inline Sources/SwiftParser/Specifiers.swift:512]: This worries me a little bit because I’m not entirely sure that we really always end up with this being `nil` but can’t come up with a counter example from the top of my head. I would feel more comfortable if we just combine it with `unexpectedBetweenAsyncSpecifierAndThrowsSpecifier`.

```swift
RawUnexpectedNodesSyntax(combining: unexpectedBetweenAsyncSpecifierAndThrowsSpecifier, unexpectedAfterThrowsSpecifier, arena: arena),
```

File: Sources/SwiftParser/Declarations.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Declarations.swift:1148]: I've copied this from `parseFunctionSignature()`. Should it be updated there as well? Could it be that this pre-check is a performance optimization? Any other reasons it might be needed there?

File: Sources/SwiftParser/Specifiers.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Specifiers.swift:486]: Done

File: Sources/SwiftParser/Specifiers.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Specifiers.swift:511]: Done

File: Sources/SwiftParser/Specifiers.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Specifiers.swift:512]: At this point we have only a single case (as noted in the extension comment), where we know that `unexpectedAfterThrowsSpecifier` cannot appear. Combining two groups of unexpected syntax looses information, which was affecting how `exchangeTokens()` behaves. Not sure if it still does after loosening the checks. But I would postpone combining or any other handling logic until other cases appear.

File: Sources/SwiftParser/Specifiers.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Specifiers.swift:486]: Actually, in the single use case where this conformance is used, `canRecoverTo()` is not used, only `at()` which seems to be perfectly fine with empty spec sets. So, I guess changes in `canRecoverTo()` are not needed. But I don't mind to do them just in case. @ahoppen, WDYT?

**Comment by nickolas-pohilets** [comment]
Rebased after https://github.com/apple/swift-syntax/pull/1767

File: Sources/SwiftParser/Specifiers.swift
Comment (ahoppen) [reply Sources/SwiftParser/Specifiers.swift:486]: I think allowing empty `TokenSpecSet` makes total sense in `canRecoverTo` and I‘d prefer to be safe here instead of discovering a stupid crash about it later if we for some reason start to use `canRecoverTo` in the future.

File: Sources/SwiftParser/Declarations.swift
Comment (ahoppen) [reply Sources/SwiftParser/Declarations.swift:1148]: If you could also update the call in `parseFunctionSignature`, that would be great. I don’t see how `self.at(.arrow)` could help performance in any significant way.

File: Sources/SwiftParser/Specifiers.swift
Comment (ahoppen) [reply Sources/SwiftParser/Specifiers.swift:511]: Did you forget to push? Still appears to be an `assert` to me.

File: Sources/SwiftParser/Specifiers.swift
Comment (ahoppen) [reply Sources/SwiftParser/Specifiers.swift:512]: I’m mostly thinking that the assumption that `unexpectedAfterThrowsSpecifier` is empty might change in the future. And especially the recovery cases are not hit as frequently in CI but more infrequently while code is being edited, so we might not actually notice that the assertion starts being violated. Thus, I’d rather be safe here. 

And I don’t see how `RawUnexpectedNodesSyntax(combining:)` can change the layout of the node if `unexpectedAfterThrowsSpecifier` is `nil` because it just grabs the elements of `unexpectedBetweenAsyncSpecifierAndThrowsSpecifier` and puts them in a new `RawUnexpectedNodesSyntax`.

Comment (nickolas-pohilets) [reply Sources/SwiftParser/Specifiers.swift]: IMO, this entire conformance is hacky, and I would advise against introducing any new cases relying on it. I think even with `RawUnexpectedNodesSyntax(combining:)` implementation of the initialiser would be not robust enough to work in the general case. It is still not possible to pass `RawDeinitEffectSpecifiersSyntax` to `parseEffectSpecifiers()`.

Cleaner solution would be create a new protocol just for the `parseMisplacedEffectSpecifiers()`. I've pushed an updated. @ahoppen, WDYT?

File: Sources/SwiftParser/Specifiers.swift
Comment (ahoppen) [inline Sources/SwiftParser/Specifiers.swift:530]: Just as I’m reading this. I would add the following comment
```
`throwsSpecifier` should never be present because `CorrectThrowsTokenKinds` is an empty `TokenSpecSet`.
```

File: Sources/SwiftParser/Declarations.swift
Comment (ahoppen) [reply Sources/SwiftParser/Declarations.swift:1148]: I think you only updated the existing call and forgot to update this one 😉

**Review by ahoppen — COMMENTED** [review]
I really like the newly introduced protocol 😍 That’s a lot cleaner than everything we had before. Thanks for the idea.

I’ve got two really minor comments, otherwise this looks good to me now. Thanks for your great collaboration on this PR.

File: Sources/SwiftParser/Specifiers.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Specifiers.swift:530]: Not even because of that, but because `parseMisplacedEffectSpecifiers()` only creates missing tokens

Comment (ahoppen) [reply Sources/SwiftParser/Specifiers.swift:530]: Ah, I see. Thanks for clearing up the misunderstanding. I think it would be good to mention that as a comment.

File: Sources/SwiftParser/Declarations.swift
Comment (nickolas-pohilets) [reply Sources/SwiftParser/Declarations.swift:1148]: 🤦‍♂️

**Comment by nickolas-pohilets** [comment]
Should this MR and https://github.com/apple/swift/pull/60057 be merged in any specific order? I was getting failing tests in the main repo without changes in this one, so I guess this MR needs to be merged first. Or neither can be merged until proposal is approved?

**Review by ahoppen — APPROVED** [review]
Nice. Thanks for your effort! One last request: Could you squash your commits? It just makes for a nicer commit history.

**Comment by nickolas-pohilets** [comment]
@swift-ci, please test

**Comment by nickolas-pohilets** [comment]
@swift-ci please test

**Comment by nickolas-pohilets** [comment]
@ahoppen, could you please trigger CI? Looks like I don’t have rights. Or what am I doing wrong?

**Comment by ahoppen** [comment]
Yes, you need push access to trigger CI for security reasons. Triggering it now.

@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by nickolas-pohilets** [comment]
CI failed because of this:
> FAIL: code-generated files committed to repository do not match generated ones. 
When I try to regenerate them, I get the following error:
```
CodeGeneration/Sources/generate-swiftsyntax/templates/swiftsyntax/KeywordFile.swift:18:48: error: type 'TokenSyntax' has no member 'leftSquareToken'
let lookupTable = ArrayExprSyntax(leftSquare: .leftSquareToken(trailingTrivia: .newline)) {
                                              ~^~~~~~~~~~~~~~~
```
Reproduces on fresh `main` branch. @ahoppen, looks like `main` is broken.

**Comment by nickolas-pohilets** [comment]
`git clean -fdx` helped. Could it be that `./build-script.py generate-source-code` incorrectly caches code generator binary?

I've pushed an update, let's re-try CI

**Comment by ahoppen** [comment]
Yes, we’re currently seeing issues with CodeGeneration not updating correctly. We haven’t had the chance to 100% pinpoint the failure point. There was a discussion going on here: https://github.com/apple/swift-syntax/pull/1741#discussion_r1229816046

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Pull request merged**
